File: worker-android/app/src/main/java/dev/sontx/logme/worker/LogMe.java
package dev.sontx.logme.worker;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.provider.Settings;
import android.text.TextUtils;
import android.util.Base64;
import android.util.Log;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import dev.sontx.logme.worker.uce.UCEDefaultActivity;
import dev.sontx.logme.worker.uce.UCEHandler;

public final class LogMe implements CommandHandler {
    private static final String TAG = LogMe.class.getName();

    private final ExecutorService executorService = Executors.newCachedThreadPool();
    private final IWorkerClient workerClient;
    private final Context context;

    public LogMe(Context context, String url) {
        this.context = context;
        new UCEHandler.Builder(context)
                .setHandleIntent(value -> {
                    String report = UCEDefaultActivity.getAllErrorDetailsFromIntent(context, value);
                    String appName = UCEDefaultActivity.getApplicationName(context);
                    String wrap = String.format("%s\n%s", appName, report);
                    send(wrap, MessageType.Exception);
                })
                .build();

        String appName = UCEDefaultActivity.getApplicationName(context).replace(" ", "");
        String clientName = getClientId(appName);
        MqttIWorkerClient mqttIWorkerClient = new MqttIWorkerClient(context, clientName, url, appName);
        mqttIWorkerClient.setCommandHandler(this);
        workerClient = mqttIWorkerClient;
    }

    private String getClientId(String appName) {
        @SuppressLint("HardwareIds")
        String androidId = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
        return String.format("%s@%s", appName, androidId);
    }

    public void start() {
        Logger.setLogMe(this);
        executorService.execute(() -> {
            try {
                workerClient.start();
            } catch (LogMeException e) {
                Log.e(TAG, "Error while starting " + getClass().getName(), e);
            }
        });
    }

    public void stop() {
        Logger.setLogMe(null);
        try {
            workerClient.stop();
        } catch (LogMeException ignored) {
        }
    }

    public void sendLog(String log) {
        send(log, MessageType.Log);
    }

    private void send(String msg, MessageType messageType) {
        executorService.execute(() -> {
            try {
                if (workerClient != null) {
                    workerClient.send(msg, messageType);
                }
            } catch (LogMeException e) {
                Log.e(TAG, "Error while sending crash logs", e);
            }
        });
    }

    @Override
    public void handleCommand(String command) {
        String data = null;
        switch (command.toUpperCase()) {
            case Constants.COMMAND_GET_SYSTEM_INFO:
                data = SystemInfo.getInfo(context);
                break;
            case Constants.COMMAND_TAKE_SCREENSHOT:
                if (context instanceof Activity) {
                    byte[] screenshot = ScreenshotHelper.takeScreenshot((Activity) context);
                    data = Base64.encodeToString(screenshot, Base64.DEFAULT);
                }
                break;
        }

        if (!TextUtils.isEmpty(data)) {
            String appName = UCEDefaultActivity.getApplicationName(context);
            String wrap = String.format("%s\n%s\n%s", command, appName, data);
            send(wrap, MessageType.ControlResponse);
        }
    }
}

File: DotNetLib/LogMe/LogMe.cs
using DeviceId;
using System;
using System.Threading.Tasks;

namespace LogMeLib
{
    public sealed class LogMe : ICommandHandler
    {
        public static async Task<LogMe> StartNewAsync(string url)
        {
            var logMe = new LogMe(url);
            await logMe.StartAsync();
            return logMe;
        }

        private readonly string url;
        private IWorkerClient workerClient;

        public LogMe(string url)
        {
            this.url = url;
        }

        public async Task StartAsync()
        {
            Logger.SetLogMe(this);

            try
            {
                if (workerClient != null)
                {
                    await workerClient.StopAsync();
                }

                var appName = SystemInfo.GetAppName();
                var clientId = GetClientId(appName);
                var mqttIWorkerClient = new MqttIWorkerClient(clientId, url, appName);
                mqttIWorkerClient.CommandHandler = this;
                await mqttIWorkerClient.StartAsync();
                workerClient = mqttIWorkerClient;
            }
            catch (Exception ex)
            {
                Logger.E("Error while starting " + GetType().Name, ex);
            }
        }

        private string GetClientId(string appName)
        {
            var deviceId = new DeviceIdBuilder()
                .AddMachineName()
                .AddMacAddress()
                .AddProcessorId()
                .AddMotherboardSerialNumber()
                .ToString();
            return $"{appName}@{deviceId}";
        }

        public async Task StopAsync()
        {
            Logger.SetLogMe(null);

            try
            {
                var workerClient = this.workerClient;
                if (workerClient != null)
                {
                    await workerClient.StopAsync();
                    this.workerClient = null;
                }
            }
            catch { }
        }

        public Task SendAsync(string log)
        {
            return SendImplAsync(log, MessageType.Log);
        }

        private async Task SendImplAsync(string msg, MessageType messageType)
        {
            try
            {
                var workerClient = this.workerClient;
                if (workerClient != null)
                {
                    await workerClient.SendAsync(msg, messageType);
                }
            }
            catch (Exception e)
            {
                Logger.E("Error while sending crash logs", e);
            }
        }

        public async void HandleCommand(string command)
        {
            string data = null;
            switch (command.ToUpper())
            {
                case Constants.COMMAND_GET_SYSTEM_INFO:
                    data = SystemInfo.GetInfo();
                    break;

                case Constants.COMMAND_TAKE_SCREENSHOT:
                    byte[] screenshot = ScreenshotHelper.TakeScreenshot();
                    data = Convert.ToBase64String(screenshot);
                    break;
            }

            if (!string.IsNullOrEmpty(data))
            {
                var appName = SystemInfo.GetAppName();
                var wrap = $"{command}\n{appName}\n{data}";
                await SendImplAsync(wrap, MessageType.ControlResponse);
            }
        }
    }
}